qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,235 @@
+            self._mixer = sum(mixer_terms)
+
+    def _construct_circuit(self, parameters) -> QuantumCircuit:
+        """Construct a parametrized circuit."""
+        if not len(parameters) == self._num_parameters:
+            raise ValueError('Incorrect number of angles: expecting {}, but {} given.'.format(
+                self._num_parameters, len(parameters)
+            ))
+
+        # local imports to avoid circular imports
+        from qiskit.opflow import CircuitStateFn
+        from qiskit.opflow import H, CircuitOp, EvolutionFactory
+
+        # initialize circuit, possibly based on given register/initial state
+        if isinstance(self._initial_state, QuantumCircuit):
+            circuit_op = CircuitStateFn(self._initial_state)
+        else:
+            circuit_op = (H ^ self._num_qubits)
+
+        # iterate over layers
+        for idx in range(self._reps):
+            # the first [:self._p] parameters are used for the cost operator,
+            # so we apply them here
+            circuit_op = (self._cost_operator * parameters[idx]).exp_i().compose(circuit_op)
+            from qiskit.opflow import OperatorBase
+            if isinstance(self._mixer, OperatorBase):
+                mixer = cast(OperatorBase, self._mixer)
+                # we apply beta parameter in case of operator based mixer.
+                circuit_op = (mixer * parameters[idx + self._reps]).exp_i().compose(circuit_op)
+            else:
+                # mixer as a quantum circuit that can be parameterized
+                mixer = cast(QuantumCircuit, self._mixer)
+                num_params = mixer.num_parameters
+                # the remaining [self._p:] parameters are used for the mixer,
+                # there may be multiple layers, so parameters are grouped by layers.
+                param_values = parameters[self._reps + num_params * idx:
+                                          self._reps + num_params * (idx + 1)]
+                param_dict = dict(zip(mixer.parameters, param_values))
+                mixer = mixer.assign_parameters(param_dict)
+                circuit_op = CircuitOp(mixer).compose(circuit_op)
+
+        evolution = EvolutionFactory.build(self._cost_operator)
+        circuit_op = evolution.convert(circuit_op)
+        return circuit_op.to_circuit()
+
+    @property
+    def cost_operator(self) -> 'OperatorBase':
+        """Returns an operator representing the cost of the optimization problem."""
+        return self._cost_operator
+
+    @cost_operator.setter
+    def cost_operator(self, cost_operator: 'OperatorBase') -> None:
+        """Sets cost operator."""
+        self._cost_operator = cost_operator
+        self._invalidate()
+
+    @property
+    def reps(self) -> int:
+        """Returns the `p` parameter, which determines the depth of the circuit."""
+        return self._reps
+
+    @reps.setter
+    def reps(self, reps: int) -> None:
+        """Sets the `p` parameter."""
```

Comment:
Was it intentional to leave 'p' here in the comments, or should this have been changed to 'reps'. (Asking as I know it corresponds to p in the referenced paper)

Reply:
Renamed.